File: 79.单词搜索.py
#
# @lc app=leetcode.cn id=79 lang=python3
#
# [79] 单词搜索
#

# @lc code=start
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        if len(board)==0:
            return False
        for i in range(len(board)):
            for j in range(len(board[0])):
                if self.backtrack(board,word,i,j,0):
                    return True
        return False

    def backtrack(self,board:List[List[str]],word:str,i:int,j:int,k:int)->bool:
        if i<0 or i>=len(board) or j<0 or j>=len(board[0]) or board[i][j]!=word[k]:
            return False
        if k==len(word)-1 and word[k]==board[i][j]:
            return True
        temp=board[i][j]
        board[i][j]='\0'
        if self.backtrack(board,word,i+1,j,k+1)\
        or self.backtrack(board,word,i,j+1,k+1)\
        or self.backtrack(board,word,i-1,j,k+1)\
        or self.backtrack(board,word,i,j-1,k+1):
            return True
        board[i][j]=temp
        return False

# @lc code=end

File: 79.单词搜索.cpp
/*
 * @lc app=leetcode.cn id=79 lang=cpp
 *
 * [79] 单词搜索
 */

// @lc code=start
class Solution
{
public:
    bool exist(vector<vector<char>> &board, string word)
    {
        if (board.size() == 0)
            return false;
        for (int i = 0; i < board.size(); i++)
            for (int j = 0; j < board[0].size(); j++)
                if (backtrack(i, j, 0, word, board))
                {
                    return true;
                }
        return false;
    }
    bool backtrack(int i, int j, int k, string &word, vector<vector<char>> &board)
    {
        if (i < 0 || i >= board.size() || j < 0 || j >= board[0].size()||board[i][j]!=word[k])
            return false;
        if (k == word.length() - 1 && word[k] == board[i][j])
            return true;
        char temp = board[i][j];
        board[i][j] = '\0';
        if (backtrack(i + 1, j, k + 1, word, board) || backtrack(i, j + 1, k + 1, word, board) || backtrack(i - 1, j, k + 1, word, board) || backtrack(i, j - 1, k + 1, word, board))
            return true;
        board[i][j] = temp;
        return false;
    }
};
// @lc code=end
